File: src/validation/semantic.ts
// validation/semantic.ts

import type {
  Workflow,
  NodeId,
  AgentDefinition,
  AgentId,
  DataType,
  PortId,
} from "../types";
import type { ValidationResult, ValidationError } from "./types";
import { validResult, invalidResult } from "./types";
import { buildReverseAdjacencyList } from "./graph";

/**
 * Registry for looking up agent definitions.
 * Passed in to avoid tight coupling to storage.
 */
export type AgentRegistry = ReadonlyMap<AgentId, AgentDefinition>;

/**
 * M1: Source output type must match target input type.
 */
export function validateTypeCompatibility(
  workflow: Workflow,
  registry: AgentRegistry,
): ValidationResult {
  const errors: ValidationError[] = [];

  for (const [edgeId, edge] of workflow.edges) {
    const sourceNode = workflow.nodes.get(edge.source);
    const targetNode = workflow.nodes.get(edge.target);

    // Skip if nodes don't exist (caught by structural validation)
    if (!sourceNode || !targetNode) continue;

    // Get agent definitions
    const sourceAgentId = sourceNode.config.agentId as AgentId | undefined;
    const targetAgentId = targetNode.config.agentId as AgentId | undefined;

    // Skip non-agent nodes (input/output nodes have dynamic types)
    if (!sourceAgentId || !targetAgentId) continue;

    const sourceAgent = registry.get(sourceAgentId);
    const targetAgent = registry.get(targetAgentId);

    if (!sourceAgent || !targetAgent) {
      errors.push({
        code: "TYPE_MISMATCH",
        message: `Unknown agent definition referenced`,
        nodeIds: [edge.source, edge.target],
        edgeIds: [edgeId],
      });
      continue;
    }

    // Find port schemas
    const sourcePort = sourceAgent.outputSchema.ports.find(
      (p) => p.name === edge.sourcePort,
    );
    const targetPort = targetAgent.inputSchema.ports.find(
      (p) => p.name === edge.targetPort,
    );

    if (!sourcePort) {
      errors.push({
        code: "TYPE_MISMATCH",
        message: `Source node has no output port: ${edge.sourcePort}`,
        nodeIds: [edge.source],
        edgeIds: [edgeId],
      });
      continue;
    }

    if (!targetPort) {
      errors.push({
        code: "TYPE_MISMATCH",
        message: `Target node has no input port: ${edge.targetPort}`,
        nodeIds: [edge.target],
        edgeIds: [edgeId],
      });
      continue;
    }

    // Check type compatibility
    if (!areTypesCompatible(sourcePort.type, targetPort.type)) {
      errors.push({
        code: "TYPE_MISMATCH",
        message: `Type mismatch: ${sourcePort.type} -> ${targetPort.type}`,
        nodeIds: [edge.source, edge.target],
        edgeIds: [edgeId],
      });
    }
  }

  return errors.length === 0 ? validResult() : invalidResult(errors);
}

/**
 * Check if source type can flow to target type.
 * Currently strict equality; can be extended for coercion rules.
 */
function areTypesCompatible(source: DataType, target: DataType): boolean {
  // Strict equality for now
  if (source === target) return true;

  // Future: could add coercion rules
  // e.g., number -> string, object -> array (if array of objects)

  return false;
}

/**
 * M2: All required inputs of a node must have incoming edges.
 */
export function validateRequiredInputs(
  workflow: Workflow,
  registry: AgentRegistry,
): ValidationResult {
  const errors: ValidationError[] = [];
  const revAdj = buildReverseAdjacencyList(workflow);

  for (const [nodeId, node] of workflow.nodes) {
    // Skip non-agent nodes
    const agentId = node.config.agentId as AgentId | undefined;
    if (!agentId) continue;

    const agent = registry.get(agentId);
    if (!agent) continue; // Caught elsewhere

    // Get incoming edges for this node
    const incomingEdgeIds = revAdj.get(nodeId) ?? [];
    const connectedPorts = new Set<string>();

    for (const edgeId of incomingEdgeIds) {
      const edge = workflow.edges.get(edgeId)!;
      connectedPorts.add(edge.targetPort as string);
    }

    // Check each required input port
    const missingPorts: string[] = [];
    for (const port of agent.inputSchema.ports) {
      if (port.required && !connectedPorts.has(port.name)) {
        missingPorts.push(port.name);
      }
    }

    if (missingPorts.length > 0) {
      errors.push({
        code: "MISSING_REQUIRED_INPUT",
        message: `Missing required inputs: ${missingPorts.join(", ")}`,
        nodeIds: [nodeId],
      });
    }
  }

  return errors.length === 0 ? validResult() : invalidResult(errors);
}

File: apps/api/src/agentforge_api/validation/semantic.py
# apps/api/src/agentforge_api/validation/semantic.py

"""
Semantic validators for DAG invariants M1-M2.

Validates type compatibility and required input satisfaction.
Requires agent registry for schema lookup.
"""

from agentforge_api.models import (
    Workflow,
    ValidationResult,
    ValidationError,
    ValidationErrorCode,
    AgentDefinition,
    DataType,
)
from agentforge_api.validation.graph import build_reverse_adjacency_list


# Type alias for agent registry
AgentRegistry = dict[str, AgentDefinition]


def are_types_compatible(source: DataType, target: DataType) -> bool:
    """
    Check if source type can flow to target type.
    Currently strict equality; can be extended for coercion rules.
    """
    # Strict equality for now
    if source == target:
        return True
    
    # Future: could add coercion rules
    # e.g., number -> string, object -> array (if array of objects)
    
    return False


def validate_type_compatibility(
    workflow: Workflow,
    registry: AgentRegistry,
) -> ValidationResult:
    """
    M1: Source output type must match target input type.
    """
    errors: list[ValidationError] = []
    node_map = workflow.get_node_map()
    
    for edge in workflow.edges:
        source_node = node_map.get(edge.source)
        target_node = node_map.get(edge.target)
        
        # Skip if nodes don't exist (caught by structural validation)
        if source_node is None or target_node is None:
            continue
        
        # Get agent IDs from node config
        source_agent_id = source_node.config.agent_id
        target_agent_id = target_node.config.agent_id
        
        # Skip non-agent nodes (input/output nodes have dynamic types)
        if source_agent_id is None or target_agent_id is None:
            continue
        
        source_agent = registry.get(source_agent_id)
        target_agent = registry.get(target_agent_id)
        
        if source_agent is None or target_agent is None:
            errors.append(ValidationError(
                code=ValidationErrorCode.TYPE_MISMATCH,
                message="Unknown agent definition referenced",
                node_ids=[edge.source, edge.target],
                edge_ids=[edge.id],
            ))
            continue
        
        # Find port schemas
        source_port = next(
            (p for p in source_agent.output_schema.ports if p.name == edge.source_port),
            None,
        )
        target_port = next(
            (p for p in target_agent.input_schema.ports if p.name == edge.target_port),
            None,
        )
        
        if source_port is None:
            errors.append(ValidationError(
                code=ValidationErrorCode.TYPE_MISMATCH,
                message=f"Source node has no output port: {edge.source_port}",
                node_ids=[edge.source],
                edge_ids=[edge.id],
            ))
            continue
        
        if target_port is None:
            errors.append(ValidationError(
                code=ValidationErrorCode.TYPE_MISMATCH,
                message=f"Target node has no input port: {edge.target_port}",
                node_ids=[edge.target],
                edge_ids=[edge.id],
            ))
            continue
        
        # Check type compatibility
        if not are_types_compatible(source_port.type, target_port.type):
            errors.append(ValidationError(
                code=ValidationErrorCode.TYPE_MISMATCH,
                message=f"Type mismatch: {source_port.type.value} -> {target_port.type.value}",
                node_ids=[edge.source, edge.target],
                edge_ids=[edge.id],
            ))
    
    if errors:
        return ValidationResult.failure(errors)
    return ValidationResult.success()


def validate_required_inputs(
    workflow: Workflow,
    registry: AgentRegistry,
) -> ValidationResult:
    """
    M2: All required inputs of a node must have incoming edges.
    """
    errors: list[ValidationError] = []
    rev_adj = build_reverse_adjacency_list(workflow)
    edge_map = workflow.get_edge_map()
    
    for node in workflow.nodes:
        # Skip non-agent nodes
        agent_id = node.config.agent_id
        if agent_id is None:
            continue
        
        agent = registry.get(agent_id)
        if agent is None:
            continue  # Caught by type compatibility validation
        
        # Get incoming edges for this node
        incoming_edge_ids = rev_adj.get(node.id, [])
        connected_ports: set[str] = set()
        
        for edge_id in incoming_edge_ids:
            edge = edge_map.get(edge_id)
            if edge:
                connected_ports.add(edge.target_port)
        
        # Check each required input port
        missing_ports: list[str] = [
            port.name
            for port in agent.input_schema.ports
            if port.required and port.name not in connected_ports
        ]
        
        if missing_ports:
            errors.append(ValidationError(
                code=ValidationErrorCode.MISSING_REQUIRED_INPUT,
                message=f"Missing required inputs: {', '.join(missing_ports)}",
                node_ids=[node.id],
            ))
    
    if errors:
        return ValidationResult.failure(errors)
    return ValidationResult.success()